Create Account Implementation

kevwu27 opened the pull request:
## Tracking Info

Resolves #14

## Changes

- Sign-up button from the Sign In page takes you to the CreateAccount page, and vice versa
- Buttons are disabled unless all fields are filled out and valid email and password length are provided
- updated button colors so the default color is the dark green, while disabled is a lighter green
- added error states
- upon clicking on SignUp button, new user is created in Firebase & new user object is stored in MongoDB
- email verification is also sent upon sign up, takes you to the EmailVerified page
- created the VerifyEmail panel 

## Testing

- comparing designs with Figma
- http://localhost:3000/emailverified/
- http://localhost:3000/signup/
- checking objects were actually created in both Firebase console and MongoDB

## Confirmation of Change

Error States & Disabled Button:
<img width="721" alt="Screenshot 2025-02-04 at 1 00 43 PM" src="https://github.com/user-attachments/assets/26168dcc-1bc8-45ff-b13c-5fd4090afefb" />

Valid Fields & Enabled Button:
<img width="720" alt="Screenshot 2025-02-04 at 1 01 52 PM" src="https://github.com/user-attachments/assets/b004da80-c2ca-4a48-93a0-fbad1943fc15" />

Verify Email Panel:
<img width="720" alt="Screenshot 2025-02-04 at 1 08 35 PM" src="https://github.com/user-attachments/assets/ea0f4799-8eb6-438c-98f8-1a618c9eb1e3" />

Email Verified Page
<img width="712" alt="Screenshot 2025-02-04 at 1 09 34 PM" src="https://github.com/user-attachments/assets/9142761e-39f4-48d5-90d0-204bcc48a789" />

km322 commented on backend/controllers/user.js:
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable */
```
Do we want to leave this in here? I think enabling lint would help with readable code? If we want to change this, make sure to delete all files with `/* eslint-disable */`

km322 commented:
Hello, your code looks amazing!

I just had a couple of questions:
1. Does the default login page have errors? For example, when you click `create account` is the image below shown? If not, I think having a default state of no errors until you start typing would be better. What do you guys think?
<img width="275" alt="Screenshot 2025-02-05 at 4 59 49 PM" src="https://github.com/user-attachments/assets/8309f0c7-6662-4e4b-bf1f-df195ebfdd00" />

2. Do the textboxes scroll? Say you have an extra long name or email, does the text box allow for that? Is there a max limit for any of the boxes?
3. Related to the question above, say the user has a really long email, does that affect the Verify Email Panel?

Other than that your code looks amazing, great job!

christen03 commented on frontend/src/app/api/user.ts:
```diff
@@ -0,0 +1,26 @@
+  try {
+    const response = await post("/api/user/signup", user);
+    const json = (await response.json()) as UserJSON;
+    console.log(json);
```
remove console log

kevwu27 replied:
resolved

christen03 commented on frontend/src/app/components/CreateAccountPanel.tsx:
```diff
@@ -1,26 +1,266 @@
+        !formData.email ||
+        !formData.password
+      ) {
+        // setError("Please fill out all fields appropriately.");
```
commented out?

christen03 commented on frontend/src/app/components/CreateAccountPanel.tsx:
```diff
@@ -1,26 +1,266 @@
+
+  useEffect(() => {
+    checkFormValidity();
+  }, [formData]);
```
do we need to checkFormValidity on every change of the form? i'm thinking we should only verify when we submit

kevwu27 replied:
it's for checking if the error messages below the text boxes should be displayed & if the sign-up should be enabled

christen03 commented on frontend/src/app/components/EmailVerifiedPanel.tsx:
```diff
@@ -0,0 +1,38 @@
+
+  useEffect(() => {
+    const emailForVerify = localStorage.getItem("emailForVerify");
+    console.log("Retrieved email from localStorage:", emailForVerify);
```
forgotten console log?

christen03 commented on frontend/src/app/components/EmailVerifiedPanel.tsx:
```diff
@@ -0,0 +1,38 @@
+      setEmail(emailForVerify);
+      localStorage.removeItem("emailForVerify");
+    } else {
+      console.error("No email found: ", emailForVerify);
```
throw error instead of console?

christen03 replied:
or show error

christen03 commented on frontend/src/app/components/ForgotPasswordForm.tsx:
```diff
@@ -1,11 +1,15 @@
+  setEmailState,
+}: {
+  setEmailState: (email: string) => void;
+}) {
```
i might be missing something but why are we pasing this in as a prop if we have the state defined below?

kevwu27 replied:
wasn't part of our PR technically as we really only changed the name of file, but it seems that the state setter in this file, setEmail, is used to keep track of the email that the user is typing

setEmailState, however, goes back to the actual ForgotPassword page, under /forgotpassword. this allows the email typed in this file (ForgotPasswordForm) to be transmitted to the ForgotPasswordEmailSent component

christen03 commented on frontend/src/app/components/SignInPanel.tsx:
```diff
@@ -1,26 +1,86 @@
+  };
+
+  const handleSubmit = () => {
+    console.log("placeholder");
```
should there be something else here?

kevwu27 replied:
will be handled in a different PR/task!

christen03 commented on frontend/src/app/components/SignInPanel.tsx:
```diff
@@ -1,26 +1,86 @@
+  };
+
+  useEffect(() => {
+    checkFormValidity();
```
same here, we should only checkValidity when we submit?

kevwu27 replied:
also for error states & checking if button should be enabled!

christen03 commented on frontend/src/app/components/VerifyEmail.tsx:
```diff
@@ -0,0 +1,58 @@
+    if (user) {
+      sendEmailVerification(user)
+        .then(() => {
+          console.log("Email sent.");
```
remove console log?

kevwu27 replied:
i think it'll help validate that firebase worked as we work on the user authentication/creation aspects, so for now we're going to keep it but before deployment it should definitely be removed

christen03 commented on frontend/src/app/signup/page.tsx:
```diff
@@ -1,16 +1,24 @@
       </section>
       <section className={styles.rightSide}>
-        <CreateAccountPanel />
+        {!isAccountCreated && <CreateAccountPanel setIsAccountCreated={setIsAccountCreated} />}
```
very very minor but we can just do the if else here with 
ternary like
{isAccountCreated ? ( </Created> ) : (<NotCreated />) } is equiv to
if(isAccountCreated){
<Created  /> 
} else {
<NotCreated />
}

christen03 reviewed (dismissed):
great job! left a few minor comments that are pretty nit picky, but loos really good!

kevwu27 commented on backend/controllers/user.js:
```diff
@@ -0,0 +1,28 @@
+      email,
+      module: 1,
+    };
+
```
in CreateAccountPanel, the email stored is at least of "correct" format meaning it contains at least a "@" and a "." after the "@". technically "@." would pass the check, but then no email would be sent through firebase 
assuming it's students who are the primary target of this application I don't think we need to worry about injection vulnerabilities for now, but i think we can keep this in mind for the future for sure

kevwu27 commented:
> Hello, your code looks amazing!
> 
> I just had a couple of questions:
> 
> 1. Do the textboxes scroll? Say you have an extra long name or email, does the text box allow for that? Is there a max limit for any of the boxes?
> 2. Related to the question above, say the user has a really long email, does that affect the Verify Email Panel?
> 
> Other than that your code looks amazing, great job!

1) yes, this was taken care of in another PR. i don't think there's a max limit as of now
2) we decided to remove the email display on the VerifyEmail panel as discussed during the meeting so that shouldnt be a problem anymore!

kevwu27 commented:
Updates:
-fixed minor bug where if you typed an insufficient length password, an error appears, but then you extend the password to exactly 6 and the error doesn't appear => was missing a dependency from the useEffect hook (in CreateAccountPanel)
-removed showing email from the EmailVerified panel
-added error message for if there is already an account with the email 

<img width="718" alt="Screenshot 2025-02-09 at 12 42 17 AM" src="https://github.com/user-attachments/assets/de8bf360-3e2c-4760-9618-903e392ce817" />

eshaan-s18 reviewed (dismissed):
Great job guys! This looks awesome and works fantastic! One super small thing I noticed was that when the account is created in Mongo, in the `name` field, there isn't a space between the first name and last name. So when I created an account the document had `name: EshaanSharma`. Could you add a space between the first name and last name when it is added to mongo? I'll approve this PR since it is a super small change. Other than that it looks awesome! Thanks for all your hard work guys 🙌
